include/triton/core/tritoncache.h

```diff
@@ -0,0 +1,261 @@
+///         - TRITONSERVER_ERROR_INTERNAL
+///           - other internal errors
+///         - nullptr
+///           - success
+TRITONCACHE_ISPEC TRITONSERVER_Error* TRITONCACHE_CacheLookup(
+    TRITONCACHE_Cache* cache, const char* key, TRITONCACHE_CacheEntry* entry);
+
+///
+/// CacheEntry Lifetime Management
+///
+
+/// Create a new cache entry object. The caller takes ownership of the
+/// TRITONCACHE_CacheEntry object and must call
+/// TRITONCACHE_CacheEntryDelete to release the object.
+///
+/// \param entry Returns the new cache entry object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryNew(
+    TRITONCACHE_CacheEntry** entry);
+
+/// Delete a cache entry object.
+///
+/// \param entry The cache entry object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryDelete(
+    TRITONCACHE_CacheEntry* entry);
+
+///
+/// CacheEntry Field Management
+///
+
+// TODO: Proper API descriptions, lifetimes, copies, etc.
+
+// Get number of items in entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemCount(
+    TRITONCACHE_CacheEntry* entry, size_t* count);
+
+// Adds item to entry
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryAddItem(
+    TRITONCACHE_CacheEntry* entry, TRITONCACHE_CacheEntryItem* item);
+
+// Gets item at index from entry where 0 <= index < count and
+// 'count' is the value returned by TRITONCACHE_CacheEntryItemCount
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryGetItem(
+    TRITONCACHE_CacheEntry* entry, size_t index,
+    TRITONCACHE_CacheEntryItem** item);
+
+
+///
+/// CacheEntryItem Lifetime Management
+///
+
+/// Create a new cache entry item object. The caller takes ownership of the
+/// TRITONCACHE_CacheEntryItem object and must call
+/// TRITONCACHE_CacheEntryItemDelete to release the object.
+///
+/// \param item Returns the new cache entry item object.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemNew(
+    TRITONCACHE_CacheEntryItem** item);
+
+/// Delete a cache entry item object.
+///
+/// \param item The cache entry item object to delete.
+/// \return a TRITONSERVER_Error indicating success or failure.
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemDelete(
+    TRITONCACHE_CacheEntryItem* item);
+
+///
+/// CacheEntryItem Field Management
+///
+
+// TODO: Proper API descriptions, lifetimes, copies, etc.
+
+// Returns number of buffers held by item
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemBufferCount(
+    TRITONCACHE_CacheEntryItem* item, size_t* count);
+
+// Adds buffer to item
+TRITONCACHE_DECLSPEC TRITONSERVER_Error* TRITONCACHE_CacheEntryItemAddBuffer(
+    TRITONCACHE_CacheEntryItem* item, const void* base, size_t byte_size);
```

Comment:
Doesn't look like we are making any copy, so the buffer owner must not release the buffer until some point. Should have a signal mechanism that the buffer owner can know the buffer can be safely release. Perhaps a callback to be provided in `TRITONCACHE_CacheEntryItemNew` and invoke in `TRITONCACHE_CacheEntryItemDelete`

Reply:
I added some comments in-line about this, please see here: https://github.com/triton-inference-server/core/pull/142/commits/4c7a7c0850df23a30c31a0939b51f16f2db67e7d

Similarly, it turned out I didn't even end up using the TRITONCACHE_CacheEntryNew/Delete APIs, and instead created the `CacheEntry`s in cache_manager and passed their opaque pointers for cache to temporarily use (get or add buffers).

I could use an extra pair of eyes/brain to help define the expected ownership, and if we even need these APIs. 

I think `TRITONCACHE_CacheEntryItemNew` makes sense to use/create an item on cache side, since we don't know the expected number of items on Triton side before calling CacheLookup. 

For `TRITONCACHE_CacheEntryItemDelete`, I don't know if it makes more sense for the cache to maintain or transfer ownership of a `CacheItem` after creating it - this would help clarify the need for signal to delete. Transferring ownership seems to work because the cache simply copies it's respective buffers (copy since cache may evict the relevant data before it's used) into the item for Triton to use.

Reply:
I think the `EntryNew/Delete` can be removed since it is not going to interfere with the cache implementation.

I would keep `TRITONCACHE_CacheEntryItemDelete` if `New` is exposed to the cache implementation, If the cache calls `New` without appending to the entry, there can be memory leak. Exposing `Delete` provides a way to clean it up in this case.

> cache simply copies it's respective buffers (copy since cache may evict the relevant data before it's used) into the item for Triton to use.

Then there are two copies? One from cache internal to cache item, another from cache item to response?

Reply:
> Then there are two copies? One from cache internal to cache item, another from cache item to response?

`TRITONCACHE_CacheLookup`:
Currently yes, one copy in `TRITONCACHE_CacheEntryItemAddBuffer()` from cache_impl -> triton_item_buffer, and then a second copy from triton_item_buffer -> response_output_buffer.

We might be able to skip the 2nd copy if there's a method to directly set an response output buffer without `AllocateDataBuffer`. I don't see one in `infer_response.cc`, but I imagine there must be some mechanism we use elsewhere.

Or we might be able to skip the first copy through some signaling mechanism where cache doesn't deallocate the buffer until cache manager copies and signals, but that might be a bit tricky.